### PR TITLE
Bubble up formatter trim warnings

### PR DESCRIPTION
**Pull request by LakshanF**
Bubbling up trim warnings to the public surface or to the binding feature boundary area seen by the [TrimTest ](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/tests/IntegrationTests/TrimTest/TrimTest.csproj)project. This is the first iteration of getting the `TrimTest `project warnings towards zero as discussed [here](https://github.com/dotnet/winforms/pull/10866#issuecomment-1940488117). Binding feature boundary is guarded by a feature check and added a RUC to `BindingContext `setter so as to generate warnings when using binding in WinForms trimming. The intention is applications like `TrimTest `that doesn't use binding will not generate warnings but applications that do use binding will get warnings at build time (when they use `BindingContext `). The runtime behavior will be a throw for trimmed WinForms applications if they use binding.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/winforms/pull/11023)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Internal/Formatter.cs [hunk 1]
```diff
@@ -263,6 +268,7 @@ internal static class Formatter
     ///  - Uses TypeConverters or IConvertible where appropriate
     ///  - Throws a FormatException if no suitable conversion can be found
     /// </summary>
+    [RequiresUnreferencedCode(ComponentModelTrimIncompatibilityMessage)]
```
[the remark this answers is not in the document]
Comment (LakshanF) [reply src/System.Windows.Forms/src/System/Windows/Forms/Internal/Formatter.cs]: `ParseObjectInternal `path is not hit at runtime. Added feature guards on the call stack path that reach the binding boundary. `FormatObjectInternal` is hit at runtime by the runtime app and bubbled up its RUC to publish methods.

**Review by KlausLoeffelmann — COMMENTED** [review]
Just had two suggestions for making messages a bit clearer - otherwise fine!

**Review by KlausLoeffelmann — DISMISSED** [review]
:shipit:

**Review by KlausLoeffelmann — APPROVED** [review]
:shipit:

**Pull request merged**